### ISSUE TITLE
Build error: The 2.2.0 SDK is required to build this repo

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#23564]
**Version Used**: bdd16d9d21a3dc59c83ef73980aa574854b32c53 (latest master)

**Steps to Reproduce**:

1. download sdk
2. run restore
3. run build

**Expected Behavior**: no error

**Actual Behavior**:

for every project:
```
\roslyn\build\Targets\Imports.targets(483,5): error : The 2.2.0 SDK is required to build this repo. It can be
install here https://dotnetcli.blob.core.windows.net/dotnet/Sdk/2.2.0-preview1-007622/dotnet-sdk-2.2.0-preview1-007622-
win-x64.exe
```

**Comment by sharwell** [comment]
Duplicate of Microsoft/msbuild#2278

**Comment by alrz** [comment]
What is the workaround? :/ this doesn't seem to work in VS nor CLI.

**Comment by jaredpar** [comment]
Are you building from a developer command prompt or a normal CMD window?

**Comment by alrz** [comment]
I tried  command prompt, developer command prompt (15.5 preview 3), cmd in git bash and PS.

I'll try with VS15.5 when it's ready.

**Comment by sharwell** [comment]
@alrz I may have misinterpreted your original steps to reproduce this issue, so reopening

**Comment by Pilchie** [comment]
Tagging @jaredpar who I think made this change recently

**Comment by jaredpar** [comment]
> I tried command prompt, developer command prompt (15.5 preview 3), cmd in git bash and PS.

Can you check and see if `%VSINSTALLDIR%` is set in the CMD / PS environment where you run Build.cmd? That can cause us to abandon our attempts at bootstrapping and merely accept the MSBuild that we are provided. 

> I'll try with VS15.5 when it's ready.

That shouldn't be necessary. Our repo is designed to self bootstrap all of our tools. Our CI actually runs on completely clean Windows 10 machines to ensure we maintain this behavior. Definitely want to work through this.

**Comment by alrz** [comment]
>Can you check and see if %VSINSTALLDIR% is set in the CMD / PS environment where you run Build.cmd? 

Yes, it's `C:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\` 

It seems like `build` tries to use the MSBuild shipped with VS.

In Rider I can choose the MSBuild file - it usually automatically detects the latest version, including MSBuild.dll shipped with sdk, I could actually build roslyn in rider using that MSBuild.dll.

>>I'll try with VS15.5 when it's ready.
>
>That shouldn't be necessary. 

Well, I did install VS15.5 and the error message is gone. it now prompts another error:

`error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.`

I do remember something like this before, I think I had to manually copy/paste sdk to where it expect it to be. (not yet tried)

**Comment by jaredpar** [comment]
> It seems like build tries to use the MSBuild shipped with VS.

Okay then that is the issue. The Build.cmd attempts to detect if it's being invoked from a developer command prompt by looking at the `%VSINSTALLDIR%` environment variable. When it detects that it will simply defer to the referenced MSBuild and make no attempt to further bootstrap tools. 

The logic behind that is that if a developer is using a VS Developer Prompt then they've made a positive assertion about what MSBuild they want to use. Hence any attempt at bootstrapping is going against their wishes.

That being said, that's also a somewhat arbitrary decision I made when I added the bootstrapping layer. I did it more out of caution than anything else. I've thought about undoing that a few times now.

**Comment by alrz** [comment]
Re: `The SDK 'Microsoft.NET.Sdk' specified could not be found.`

I uninstalled all core sdks versions and let `build.cmd` download it itself. but it fails (my connection is slow) -- And when I install it manually still it tries to download it. I think it had trouble locating the sdk (now) or confused multiple versions (before).

**Comment by alrz** [comment]
where else this depends on that might be the source of problem? should reinstall windows?

**Comment by alrz** [comment]
It would be nice if I could download sdk zip manually so build doesn't have to download it.

**Comment by jaredpar** [comment]
> I think it had trouble locating the sdk (now) or confused multiple versions (before).

It's still just keying off of `%VSINSTALLDIR%`. Can you delete that environment variable and re-run Build.cmd?

**Comment by alrz** [comment]
it's only defined on developer command prompt.

command prompt:
```
D:\dotnet\roslyn>echo %VSINSTALLDIR%
%VSINSTALLDIR%

D:\dotnet\roslyn>build
Repo Dir D:\dotnet\roslyn
Binaries Dir D:\dotnet\roslyn\Binaries
Downloading CLI 2.2.0-preview1-007622
```


developer command prompt:
```
D:\dotnet\roslyn>echo %VSINSTALLDIR%
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\

D:\dotnet\roslyn>build
Repo Dir D:\dotnet\roslyn
Binaries Dir D:\dotnet\roslyn\Binaries
Downloading CLI 2.2.0-preview1-007622
```
both:
```cs
D:\dotnet\roslyn>dotnet --version
2.2.0-preview1-007622
```

**Comment by alrz** [comment]
Ok, I had to change `dotnetRuntimeVersion` to `2.1.0-preview1-25929-02`

**Comment by alrz** [comment]
@jaredpar FYI This came back.

 (regular cmd - no %VSINSTALLDIR%)
```
...
D:\dotnet\roslyn\build\Targets\Imports.targets(481,5): error : The 2.0 SDK is required to build this repo [D:\dotnet\ro
slyn\src\Tools\RoslynPublish\RoslynPublish.csproj]
D:\dotnet\roslyn\build\Targets\Imports.targets(481,5): error : The 2.0 SDK is required to build this repo [D:\dotnet\ro
slyn\src\CodeStyle\CSharp\Tests\CSharpCodeStyleTests.csproj]
D:\dotnet\roslyn\build\Targets\Imports.targets(481,5): error : The 2.0 SDK is required to build this repo [D:\dotnet\ro
slyn\src\CodeStyle\VisualBasic\Tests\BasicCodeStyleTests.vbproj]
D:\dotnet\roslyn\build\Targets\Imports.targets(481,5): error : The 2.0 SDK is required to build this repo [D:\dotnet\ro
slyn\src\CodeStyle\VisualBasic\Tests\BasicCodeStyleTests.vbproj]
    0 Warning(s)
    157 Error(s)

Time Elapsed 00:00:02.16
Command failed to execute: C:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\MSBuild\15.0\Bin\msbuild.exe /p:TreatWarningsAsErrors=true /warnaserror /nologo /nodeReuse:false /consoleloggerparameters:Verbosity=minimal;summary /p:Configuration=Debug /m Roslyn.sln /p:DeployExtension=false
System.Management.Automation.RuntimeException: Command failed to execute: C:\Program Files (x86)\Microsoft Visual Studio\Preview\Community\MSBuild\15.0\Bin\msbuild.exe /p:TreatWarningsAsErrors=true /warnaserror /nologo /nodeReuse:false /consoleloggerparameters:Verbosity=minimal;summary /p:Configuration=Debug /m Roslyn.sln /p:DeployExtension=false
at Exec-CommandCore, D:\dotnet\roslyn\build\scripts\build-utils.ps1: line 68
at Exec-Console, D:\dotnet\roslyn\build\scripts\build-utils.ps1: line 100
at Run-MSBuild, D:\dotnet\roslyn\build\scripts\build.ps1: line 145
at Build-Artifacts, D:\dotnet\roslyn\build\scripts\build.ps1: line 170
at <ScriptBlock>, D:\dotnet\roslyn\build\scripts\build.ps1: line 579
```

**Comment by alrz** [comment]
This is log from msbuild

![image](https://user-images.githubusercontent.com/3105979/33859243-3a758678-dee8-11e7-94b6-a62f47feeae5.png)

Actual version of the msbuild
```
C:\...>msbuild
Microsoft (R) Build Engine version 15.5.180.51428 for .NET Framework
```

**Comment by jaredpar** [comment]
Can you attach the full output of your build.cmd execution somewhere I can look at? 

Side note: have you tried repairing your VS installation? The scripts here are pretty aggressive about ensuring the correct SDK is either a) installed or b) installed for you during Build.cmd. This is the only machine we've seen reliable failures on. I'm wondering if there is something broken in the installation here. That's a reach though.

**Comment by alrz** [comment]
This is the full output. [MSBuildStructuredLog-15568.zip](https://github.com/dotnet/roslyn/files/1552785/MSBuildStructuredLog-15568.zip)


I'll try repairing VS whenever I could. Thanks.

**Comment by alrz** [comment]
Somehow I managed to get a build with `dotnet build` in git bash but not cmd. There is definitely something wrong with my machine config. I'll try to figure it out if it fails again. you can dismiss the issue.

Just one thing that I observed earlier https://github.com/dotnet/roslyn/issues/23564#issuecomment-349334185. Sounds like if you have the right sdk but the wrong runtime the error message still complains about the sdk.
